lib/decoder/real-time.js

```diff
@@ -0,0 +1,66 @@
+const url = require('url')
+
+// https://github.com/PRX/Infrastructure/blob/master/cdn/dovetail3-cdn.yml#L289
+const FIELDS = [
+  'timestamp',
+  'c-ip',
+  'sc-status',
+  'cs-method',
+  'cs-uri-stem',
+  'cs-user-agent',
+  'cs-referer',
+  'x-forwarded-for',
+  'sc-content-len',
+  'sc-range-start',
+  'sc-range-end',
+]
+const ASCII = /^[\x00-\xFF]*$/
+
+/**
+ * check for cloudfront real-time log events, with known TSV fields
+ */
+exports.detect = function (buffer) {
+  const str = buffer.toString('utf-8')
+  return ASCII.test(str) && str.split('\t').length === FIELDS.length
+}
+
+/**
+ * parse tab separated log lines
+ */
+exports.decode = async function (buffer) {
+  const parts = buffer.toString('utf-8').trim().split('\t')
+  const data = {}
+  FIELDS.forEach((name, index) => {
+    if (parts[index] === '-') {
+      data[name] = null
+    } else {
+      data[name] = parts[index]
+    }
+  })
+
+  // only 2XX GET requests count
+  if (data['cs-method'] !== 'GET' || data['sc-status'] < '200' || data['sc-status'] >= '300') {
+    return null
+  }
+
+  // couple things to parse out
+  const time = Math.round(parseFloat(data.timestamp) * 1000)
+  const le = url.parse(data['cs-uri-stem'], true).query.le
+  const digest = (data['cs-uri-stem'] || '').split('/').slice(-2, -1)[0]
+  if (!time || !le || !digest) {
+    return null
+  }
+
+  // calculate actual byte range returned
+  const length = parseInt(data['sc-content-len'], 10) || 0
+  const start = parseInt(data['sc-range-start'], 10) || 0
+  const end = parseInt(data['sc-range-end'], 10) || length - 1
```

Comment:
Not quite sure how these values get used, but I'd think this needs to be `(length - 1 + start)`? Otherwise you could have `start=100` and `end=2` if `length=3`. I'd expect `100 to 102`

Reply:
Hrmm... I'm pretty sure if you ask for `Range: bytes=100-`, the logs will still have both `sc-range-start` and `sc-range-end` set.  But you're right, that's a good sanity check.